docs/spec/spec.md

```diff
@@ -2758,9 +2763,215 @@ public function main() returns error? {
+    resolveReference: resolveProduct
+}
+type Product record {
+    string id;
+    string name;
+    int price;
+};
+```
+            </pre>
+        </td>
+        <td>
+            <pre lang="graphql">
+```graphql
+type Product @key(fields: "id") @key(fields: "sku") {
+    id: String!
+    name: String!
+    price: Int!
+}
+```
+            </pre>
+        </td>
+    </tr>
+    <tr>
+        <td>
+            Compound key
+        </td>
+        <td>
+            <pre lang="ballerina">
+```ballerina
+@subgraph:Entity {
+    key: "id organization { id }",
+    resolveReference: resolveUser
+}
+type User record {
+    string id;
+    Organization organization;
+}
+```
+            </pre>
+        </td>
+        <td>
+            <pre lang="graphql">
+```graphql
+type User @key(fields: "id organization { id }") {
+    id: String!
+    organization: Organization!
+}
+```
+            </pre>
+        </td>
+    </tr>
+      <tr>
+        <td>
+            Non resolvable
+        </td>
+        <td>
+            <pre lang="ballerina">
+```ballerina
+@subgraph:Entity {
+    key: "id"
+    resolveReference: ()
+}
+type Product record {
+  id: String!
+}
+```
+            </pre>
+        </td>
+        <td>
+            <pre lang="graphql">
+```graphql
+type Product @key(fields: "id", resolvable: false) {
+    id: ID!
+}
+```
+            </pre>
+        </td>
+    </tr>
+</table>
+
+#### 13.1.3 The `subgraph:ReferenceResolver`
+
+Reference resolver is a function that resolves an entity of a specific type using its primary key. When the graph router needs to resolve an entity, it calls the reference resolver and passes the primary key and the `__typename` field of the entity. The reference resolver then returns the entity with the given primary key. Following is the type definition of a reference resolver defined in `graphql.subgraph` module.
+
+```ballerina
+public type ReferenceResolver function (subgraph:Representation representation) returns record {}|service object {}|error?;
+```
+Here, `subgraph:Representation` is a type definition of the entity representation outlined in the federation specification, which includes the GraphQL `__typename` field of the entity being resolved and its primary key.
+
+###### Example: A Product Entity Defined with Its Resolver
+
+```ballerina
+function resolveProduct(subgraph:Representation representation) returns Product|error? {
+    string id = check representation["id"].ensureType(); // obtain the primary key of the entity
+    return findProduct(id);
+}
+
+@subgraph:Entity {
+    key: "id", // primary key of the entity
+    resolveReference: resolveProduct
+}
+type Product record {
+    string id;
+    string name;
+    int price;
+};
+```
+
+>**Note:** If the reference resolver returns an entity of a different type than the entity being resolved, a runtime error will be returned to the router. For example, if the resolver returns a `User` for a `Product` entity, a runtime error will occur.
```

Comment:
I'm wondering whether we can validate this in the compile-time. Given a reference resolver, can't we validate its return type at the compile-time?

Reply:
We should be able to validated this up-to certain level by walking the syntax tree. This is the previously created issue to track this: https://github.com/ballerina-platform/ballerina-standard-library/issues/4144